src/main/java/com/google/json/JsonSanitizer.java

```diff
@@ -541,16 +541,47 @@ private void sanitizeString(int start, int end) {
             }
           }
           break;
-        // Embedding.  Disallow </script and ]]> in string literals so that
-        // the output can be embedded in HTML script elements and in XML CDATA
-        // sections.
+        // Embedding. Disallow <script, </script, <!--, --> and ]]> in string
+        // literals so that the output can be embedded in HTML script elements
+        // and in XML CDATA sections without affecting the parser state.
+        // References:
+        // https://www.w3.org/TR/html53/semantics-scripting.html#restrictions-for-contents-of-script-elements
+        // https://www.w3.org/TR/html53/syntax.html#script-data-escaped-state
+        // https://www.w3.org/TR/html53/syntax.html#script-data-double-escaped-state
+        // https://www.w3.org/TR/xml/#sec-cdata-sect
+        case 'S':
+        case 's':
+          // Disallow <script, which followed by various characters lets the
+          // HTML parser switch into or out of the "script data double escaped"
+          // state.
+          if (i > start && i + 1 < end && '<' == jsonish.charAt(i - 1)
+              && 'c' == (jsonish.charAt(i + 1) | 32)) {
+            replace(i, i + 1, (ch == 'S') ? "\\u0053" : "\\u0073"); // Escaped S or s
+          }
+          break;
         case '/':
-          // Don't over escape.  Many JSON bodies contain innocuous HTML
-          // that can be safely embedded.
+          // Disallow </script, which ends a script block.
           if (i > start && i + 2 < end && '<' == jsonish.charAt(i - 1)
               && 's' == (jsonish.charAt(i + 1) | 32)
               && 'c' == (jsonish.charAt(i + 2) | 32)) {
-            insert(i, '\\');
+            insert(i, '\\'); // Escape the /
+          }
+          break;
+        case '>':
+          // Disallow -->, which lets the HTML parser switch out of the "script
+          // data escaped" or "script data double escaped" state.
+          if (i > start + 1 && '-' == jsonish.charAt(i - 2)
+              && '-' == jsonish.charAt(i - 1)) {
+            replace(i, i + 1, "\\u003e"); // Escaped >
+          }
+          break;
+        case '!':
```

Comment:
Instead of looking for *s*'s above, Could we instead escape '<' when followed by one of the two character sequences

- "!-"
- "sc" any capitalization
- "/s", any capitalization

That should get us where we need without blowing up the output size when encoding HTML.

Something like

    // bounds check
    char c1 = jsonish.charAt(i + 1);
    char c2 = jsonish.charAt(i + 2);
    char lc1 = c1 | 32;
    char lc2 = c2 | 32;
    if (c1 == '!' && c2 == '-' || lc1 == 's' && lc2 == 'c' || c1 == '/' && lc2 == 's') {
      ...
    }

I might dig a little and see if there are Unicode normal forms that decompose to include those pairs but there's no need to delay this CL for that.

Reply:
I force-pushed a new version where I follow this approach, but I decided to check for three characters since that prevents escaping for standard HTML end tags that begin with `</s`.